syntax = "proto3";

package khaleesi.core.sawmill;

import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";
import "core.proto";


/* SHARED *****************************************************************************************/


message ResponseMetadata {
  google.protobuf.Timestamp loggedTimestamp = 1;
  string                    errors          = 2;
}


message Event {

  message Target {
    string             type  = 1;
    string             id    = 2;
    khaleesi.core.User owner = 3;
  }

  message Action {
    enum ActionType {
      // Special.
      UNKNOWN_ACTION = 0;
      CUSTOM         = 1;

      // CRUD.
      CREATE = 10;
      READ   = 11;
      UPDATE = 12;
      DELETE = 13;

      // Processes.
      START = 20;
      END   = 21;
    }

    enum ResultType {
      UNKNOWN_RESULT = 0;
      SUCCESS        = 1;
      WARNING        = 2;
      ERROR          = 3;
      FATAL          = 4;
    }

    ActionType crudType   = 1;
    string     customType = 2;
    ResultType result     = 3;
    string     details    = 4;
  }

  khaleesi.core.RequestMetadata requestMetadata = 1;
  string                        id              = 2;
  Target                        target          = 3;
  Action                        action          = 4;

  bool loggerSendMetric = 10;
}


message HttpRequest {
  // Metadata.
  khaleesi.core.RequestMetadata requestMetadata = 1;

  // khaleesi.ninja.
  string language = 10;
  string deviceId = 11;

  // HTTP.
  string languageHeader = 20;
  string ip             = 21;
  string useragent      = 22;
}

message GrpcRequest {
  khaleesi.core.RequestMetadata   requestMetadata = 1;
  khaleesi.core.GrpcCallerDetails upstreamRequest = 2;
}

message Query {
  string id         = 1;
  string raw        = 2;
  string connection = 3;

  google.protobuf.Timestamp start = 10;
  google.protobuf.Timestamp end   = 11;
}

message Error {
  khaleesi.core.RequestMetadata requestMetadata = 1;
  string                        id              = 2;
  string                        status          = 3;
  string                        loglevel        = 4;
  string                        gate            = 5;
  string                        service         = 6;
  string                        publicKey       = 7;
  string                        publicDetails   = 8;
  string                        privateMessage  = 9;
  string                        privateDetails  = 10;
  string                        stacktrace      = 11;
}


message Response {
  string                    status    = 1;
  google.protobuf.Timestamp timestamp = 2;
}

message HttpResponseRequest {
  khaleesi.core.RequestMetadata requestMetadata = 1;
  Response                      response         = 2;
}

message GrpcResponseRequest {
  khaleesi.core.RequestMetadata requestMetadata = 1;
  Response                      response        = 2;
  repeated Query                queries         = 3;
}


/* LUMBERJACK *************************************************************************************/


message LogStandardResponse {}

service Lumberjack {
  rpc LogEvent               (Event)               returns (LogStandardResponse);
  rpc LogGrpcRequest         (GrpcRequest)         returns (LogStandardResponse);
  rpc LogGrpcResponse        (GrpcResponseRequest) returns (LogStandardResponse);
  rpc LogError               (Error)               returns (LogStandardResponse);
  rpc LogHttpRequest         (HttpRequest)         returns (LogStandardResponse);
  rpc LogHttpRequestResponse (HttpResponseRequest) returns (LogStandardResponse);
  rpc LogSystemHttpRequest   (khaleesi.core.EmptyRequest) returns (LogStandardResponse);
}


message EventResponse {
  ResponseMetadata eventMetadata = 1;
  Event            event         = 2;
}

message EventsList {
  repeated EventResponse events = 1;
}

message QueryResponse {
  khaleesi.core.RequestMetadata queryRequestMetadata  = 1;
  ResponseMetadata              queryResponseMetadata = 2;
  Query                         query                 = 3;

  string                   normalized       = 10;
  repeated string          tables           = 11;
  repeated string          columns          = 12;
  google.protobuf.Duration reportedDuration = 20;

}

message QueryList{
  repeated QueryResponse queries = 1;
}


message ProcessedResponse {
  google.protobuf.Duration reportedDuration = 1;
  google.protobuf.Duration loggedDuration   = 2;
  google.protobuf.Duration childDurationAbsolute = 10;
  double                   childDurationRelative = 11;
}


message HttpRequestResponse {
  ResponseMetadata requestMetadata  = 1;
  HttpRequest      request          = 2;
  ResponseMetadata responseMetadata = 3;
  Response         response         = 4;

  enum RequestType {
    UNKNOWN = 0;
    USER    = 1;
    SYSTEM  = 2;
  }

  // Processed data.
  RequestType       type              = 10;
  ProcessedResponse processedResponse = 11;

  string geolocation    = 20;
  string browser        = 21;
  string renderingAgent = 22;
  string os             = 23;
  string deviceType     = 24;
}

message HttpRequestList {
  repeated HttpRequestResponse requests = 1;
}


message GrpcRequestResponse {
  ResponseMetadata  requestMetadata   = 1;
  GrpcRequest       request           = 2;
  ResponseMetadata  responseMetadata  = 3;
  Response          response          = 4;
  ProcessedResponse processedResponse = 5;
}

message GrpcRequestList {
  repeated GrpcRequestResponse requests = 1;
}

message ErrorResponse {
  ResponseMetadata errorMetadata = 1;
  Error            error         = 2;
}

message ErrorList {
  repeated ErrorResponse errors = 1;
}


message LogFilter {
  khaleesi.core.RequestMetadata requestMetadata = 1;
}


service Sawyer {
  rpc GetEvents           (LogFilter) returns (EventsList);
  rpc GetGrpcRequests     (LogFilter) returns (GrpcRequestList);
  rpc GetErrors           (LogFilter) returns (ErrorList);
  rpc GetHttpRequests     (LogFilter) returns (HttpRequestList);
  rpc GetQueries          (LogFilter) returns (QueryList);
  rpc CleanupEvents       (khaleesi.core.JobExecutionRequest) returns (khaleesi.core.EmptyResponse);
  rpc CleanupGrpcRequests (khaleesi.core.JobExecutionRequest) returns (khaleesi.core.EmptyResponse);
  rpc CleanupErrors       (khaleesi.core.JobExecutionRequest) returns (khaleesi.core.EmptyResponse);
  rpc CleanupHttpRequests (khaleesi.core.JobExecutionRequest) returns (khaleesi.core.EmptyResponse);
  rpc CleanupQueries      (khaleesi.core.JobExecutionRequest) returns (khaleesi.core.EmptyResponse);
}


/* FORESTER ***************************************************************************************/


message CallData {
  khaleesi.core.GrpcCallerDetails          call     = 1;
  repeated khaleesi.core.GrpcCallerDetails calls    = 2;
  repeated khaleesi.core.GrpcCallerDetails calledBy = 3;
}

message ServiceCallData {
  repeated CallData callList = 1;
}

service Forester {
  rpc GetServiceCallData (khaleesi.core.EmptyRequest) returns (ServiceCallData);
}
